service/frontend/adminHandler.go

```
@@ -654,50 +660,113 @@ func (adh *AdminHandler) DescribeCluster(_ context.Context, _ *adminservice.Desc
-	  4. persist the remote cluster metadata
-	*/
-	return nil, adh.error(serviceerror.NewUnimplemented("TODO"), scope)
+	adminClient, err := adh.Resource.GetClientFactory().NewAdminClientWithTimeout(
+		request.GetFrontendAddress(),
+		admin.DefaultTimeout,
+		admin.DefaultLargeTimeout,
+	)
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	// Fetch cluster metadata from remote cluster
+	resp, err := adminClient.DescribeCluster(ctx, &adminservice.DescribeClusterRequest{})
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	err = adh.validateRemoteClusterMetadata(resp)
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	clusterMetadataMrg := adh.GetClusterMetadataManager()
+	clusterData, err := clusterMetadataMrg.GetClusterMetadata(
+		&persistence.GetClusterMetadataRequest{ClusterName: resp.GetClusterName()},
+	)
+	if _, notFound := err.(*serviceerror.NotFound); err != nil && !notFound {
+		return nil, adh.error(err, scope)
+	}
+	var updateRequestVersion int64 = 0
+	if clusterData != nil {
+		updateRequestVersion = clusterData.Version
+	}
+
+	applied, err := clusterMetadataMrg.SaveClusterMetadata(&persistence.SaveClusterMetadataRequest{
+		ClusterMetadata: persistencespb.ClusterMetadata{
+			ClusterName:              resp.GetClusterName(),
+			HistoryShardCount:        resp.GetHistoryShardCount(),
+			ClusterId:                resp.GetClusterId(),
+			ClusterAddress:           request.GetFrontendAddress(),
+			FailoverVersionIncrement: resp.GetFailoverVersionIncrement(),
+			InitialFailoverVersion:   resp.GetInitialFailoverVersion(),
+			IsGlobalNamespaceEnabled: resp.GetIsGlobalNamespaceEnabled(),
+			IsConnectionEnabled:      request.GetEnableRemoteClusterConnection(),
+		},
+		Version: updateRequestVersion,
+	})
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+	if !applied {
+		return nil, adh.error(serviceerror.NewInvalidArgument(
+			"Cannot add remote cluster due to update immutable fields"),
+			scope,
+		)
```

would great to add which field violate the immutability

---

This info would be good to pop up from the cluster metadata store. It just return a bool right now. We can return an error to have this information. However, I am not sure if this would be a API behavior change. I will take a look and see if I can do it in a separate PR.